### DISCUSSION TITLE
Typescript vs tsParticles

### DISCUSSION THREAD

**Discussion by Orenjiku** [matteobruni/tsparticles#4081] [Q&A] [answered]
I've been messing around with tsParticles in React and so far it's been great. My only problem is right now I have the App component as a .jsx file because I get type errors for main and container objects. When I set those to any type, I get errors on the Particles component "loaded" and "options" attributes. Is there a type module I can install to handle these errors?

**Comment by matteobruni** [comment, marked as answer]
> I've been messing around with tsParticles in React and so far it's been great. My only problem is right now I have the App component as a .jsx file because I get type errors for main and container objects. When I set those to any type, I get errors on the Particles component "loaded" and "options" attributes. Is there a type module I can install to handle these errors?

Those types are available in `tsparticles-engine`, you can see a sample in the `create-react-app` templates here: https://github.com/matteobruni/tsparticles/blob/main/templates/react-ts/template/src/App.tsx

**Reply by Orenjiku** [reply]
I followed the example. Now I have a single error on the Particles "init" attribute.  
```
Type '(engine: Engine) => void' is not assignable to type '(engine: Engine) => Promise<void>'.
  Type 'void' is not assignable to type 'Promise<void>'.ts(2322)
```
Below is my component. I checked over the spelling and everything.
```
import React, { useCallback } from 'react'
import Particles from 'react-tsparticles'
import { loadFull } from 'tsparticles'
import type { Engine } from 'tsparticles-engine'
import { ISourceOptions } from 'tsparticles-engine'
import { particlesOptions } from '../particlesOptions/particlesOptionsSmoke'

const ParticlesCanvas = () => {
  const particlesInit = useCallback((engine: Engine) => {
    loadFull(engine)
  }, [])

  return (
    <Particles
      options={particlesOptions as ISourceOptions}
      init={particlesInit}
    />
  )
}

export default ParticlesCanvas
```

I also used the create-react-app template just to compare and the error shows up there too.
```
npx create-react-app my-app --template particles-typescript  
```

**Reply by matteobruni** [reply]
> I followed the example. Now I have a single error on the Particles "init" attribute.
> 
> ```
> Type '(engine: Engine) => void' is not assignable to type '(engine: Engine) => Promise<void>'.
>   Type 'void' is not assignable to type 'Promise<void>'.ts(2322)
> ```
> 
> Below is my component. I checked over the spelling and everything.
> 
> ```
> import React, { useCallback } from 'react'
> import Particles from 'react-tsparticles'
> import { loadFull } from 'tsparticles'
> import type { Engine } from 'tsparticles-engine'
> import { ISourceOptions } from 'tsparticles-engine'
> import { particlesOptions } from '../particlesOptions/particlesOptionsSmoke'
> 
> const ParticlesCanvas = () => {
>   const particlesInit = useCallback((engine: Engine) => {
>     loadFull(engine)
>   }, [])
> 
>   return (
>     <Particles
>       options={particlesOptions as ISourceOptions}
>       init={particlesInit}
>     />
>   )
> }
> 
> export default ParticlesCanvas
> ```
> 
> I also used the create-react-app template just to compare and the error shows up there too.
> 
> ```
> npx create-react-app my-app --template particles-typescript  
> ```

My bad, the callback needs to be async, I'll fix it in the template soon, I didn't notice before, thanks for the report

``` typescript
    const particlesInit = useCallback(async (engine: Engine) => {
        await loadFull(engine);
    }, []);
```

This is how it should be

**Reply by Orenjiku** [reply]
Thanks that fixed it. Also, is there a discord or somewhere to ask other questions about some of the features? I was able to test out quite a bit of the features and figure out what they do, but some features I'm not sure of. Also, is it possible to add the grab property to the onClick event under interactivity, and add color to interactivity.modes.connect.links? The color of the interactivity.modes.connect.links seems to be matched to the particles.color, but I think it would be nice to have the color independent.

**Reply by matteobruni** [reply]
You can find all links in this repository readme or in the website https://particles.js.org/

**Reply by Orenjiku** [reply]
Thanks. Really appreciate your help and enjoying what you built.
